src/runtime/Codecs/FunctionCodec.cs

```diff
@@ -0,0 +1,377 @@
+                return IntPtr.Zero;
+            }
+
+            return Runtime.PyNone;
+        }
+    }
+
+    internal class FunctionWrapper : ICallable
+    {
+        Func<object[], object> _func = null;
+
+        internal FunctionWrapper(Func<object> func)
+        {
+            _func = (object[] args) => { return func(); };
+        }
+
+        internal FunctionWrapper(Func<object[], object> func)
+        {
+            _func = func;
+        }
+
+        public virtual IntPtr RunAction(IntPtr self, IntPtr args)
+        {
+            try
+            {
+                object[] managedArgs = CallableHelper.ConvertArgs(args);
+                var retVal = _func(managedArgs);
+                return Converter.ToPython(retVal);
+            }
+            catch
+            {
+                Exceptions.SetError(Exceptions.TypeError, "Error in RunAction");
+                return IntPtr.Zero;
+            }
+        }
+    }
+
+    //converts python functions to C# actions
+    class FunctionCodec : IPyObjectDecoder, IPyObjectEncoder
+    {
+        private static int GetNumArgs(PyObject pyCallable)
+        {
+            var locals = new PyDict();
+            locals.SetItem("f", pyCallable);
+            using (Py.GIL())
+                PythonEngine.Exec(@"
+from inspect import signature
+x = len(signature(f).parameters)
+", null, locals.Handle);
+
+            var x = locals.GetItem("x");
+            return new PyInt(x).ToInt32();
+        }
+
+        private static bool IsUnaryAction(Type targetType)
+        {
+            return targetType == typeof(Action);
+        }
+
+        private static bool IsVariadicObjectAction(Type targetType)
+        {
+            return targetType == typeof(Action<object[]>);
+        }
+
+        private static bool IsUnaryFunc(Type targetType)
+        {
+            return targetType == typeof(Func<object>);
+        }
+
+        private static bool IsVariadicObjectFunc(Type targetType)
+        {
+            return targetType == typeof(Func<object[], object>);
+        }
+
+        private static bool IsAction(Type targetType)
+        {
+            return IsUnaryAction(targetType) || IsVariadicObjectAction(targetType);
+        }
+
+        private static bool IsFunc(Type targetType)
+        {
+            return IsUnaryFunc(targetType) || IsVariadicObjectFunc(targetType);
+        }
+
+        private static bool IsCallable(Type targetType)
+        {
+            //TODO - delegate, event, ...
+            return IsAction(targetType) || IsFunc(targetType);
+        }
+
+        public static FunctionCodec Instance { get; } = new FunctionCodec();
+        public bool CanDecode(PyObject objectType, Type targetType)
+        {
+            //python object must be callable
+            if (!objectType.IsCallable()) return false;
+
+            //C# object must be an Action
+            if (!IsCallable(targetType))
+                return false;
```

Comment:
or a function?
Is there a reason you can't support arbitrary delegate type?

Reply:
@lostmsu the main reason is my ignorance of C# delegate types